tools/otci/otci/otci.py

```diff
@@ -2619,6 +2638,165 @@ def wait_for(self, command: str, expect_line: Optional[Union[str, Pattern, Colle
+        output = self.execute_platform_command(cmd, timeout=transmit_time + wait_time)
+
+        results = {}
+        for line in output:
+            fields = line.split()
+            if len(fields) == 13:
+                if fields[-1] == 'sender':
+                    results['sender'] = self.__parse_iperf3_report(line)
+                elif fields[-1] == 'receiver':
+                    results['receiver'] = self.__parse_iperf3_report(line)
+
+        return results
+
+    def iperf3_server(self,
+                      bind_address: Optional[Union[str, Ip6Addr]] = None,
+                      interval: int = 1,
+                      timeout: int = 60) -> Dict[str, Any]:
+        """Run iperf3 in server mode.
+
+        :param bind_address: The local address to be bound.
+        :param interval: Seconds between periodic throughput reports (default 1 sec).
+        :param timeout: Timeout in seconds to abort the program (default 60 secs)
+        """
+        #
+        # Iperf3 server example:
+        #
+        # $ iperf3 -s -1 -B fd6e:debe:d06c:f110:7903:1c0a:854d:a991
+        # -----------------------------------------------------------
+        # Server listening on 5201
+        # -----------------------------------------------------------
+        # Accepted connection from fdb5:1c54:4233:1:f316:192e:626b:171f, port 57252
+        # [  5] local fd6e:debe:d06c:f110:7903:1c0a:854d:a991 port 5201 connected to fdb5:1c54:4233:1:f316:192e:626b:171f port 33577
+        # [ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
+        # [  5]   0.00-1.00   sec  4.81 KBytes  39.4 Kbits/sec  16.729 ms  0/4 (0%)
+        # [  5]   1.00-2.00   sec  6.02 KBytes  49.3 Kbits/sec  34.896 ms  0/5 (0%)
+        # [  5]   2.00-3.00   sec  3.61 KBytes  29.6 Kbits/sec  64.739 ms  0/3 (0%)
+        # [  5]   3.00-4.00   sec  4.81 KBytes  39.4 Kbits/sec  74.125 ms  0/4 (0%)
+        # [  5]   4.00-5.00   sec  6.02 KBytes  49.3 Kbits/sec  95.799 ms  8/13 (62%)
+        # [  5]   5.00-6.00   sec  7.22 KBytes  59.1 Kbits/sec  80.093 ms  6/12 (50%)
+        # [  5]   6.00-6.35   sec  2.41 KBytes  55.7 Kbits/sec  80.014 ms  2/4 (50%)
+        # - - - - - - - - - - - - - - - - - - - - - - - - -
+        # [ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
+        # [  5]   0.00-6.35   sec  34.9 KBytes  45.0 Kbits/sec  80.014 ms  16/45 (36%)  receiver
+
+        bind_option = '' if bind_address is None else f'-B {bind_address}'
+        interval_option = f'-i {interval}'
+
+        cmd = f'iperf3 -s -1 {bind_option} {interval}'
+        output = self.execute_platform_command(cmd, timeout)
+
+        results = {}
+        for line in output:
+            fields = line.split()
+            if len(fields) == 13 and fields[-1] == 'receiver':
+                results = self.__parse_iperf3_report(line)
+
+        return results
+
+    def __parse_iperf3_report(self, line: str) -> Dict[str, Any]:
```

Comment:
Have you tried the `--json` option? Will it simplify the parsing work?

Reply:
Yes, I tried the `--json` option. I found that both the iperf3 client and server reports do not contain the receiver's `bits_per_second`. So that we can't use the `--json` option to get the throughput.
iperf3 client:

```
   ...
   "end":  {
           "streams":      [{
                           "udp":  {
                                   "socket":       5,
                                   "start":        0,
                                   "end":  10.000154,
                                   "seconds":      10.000154,
                                   "bytes":        113344,
                                   "bits_per_second":      90673.8036234242,
                                   "jitter_ms":    38.299352718553145,
                                   "lost_packets": 10,
                                   "packets":      92,
                                   "lost_percent": 10.869565217391305,
                                   "out_of_order": 0,
                                   "sender":       true
                           }
                   }],
           "sum":  {
                   "start":        0,
                   "end":  10.871231,
                   "seconds":      10.871231,
                   "bytes":        113344,
                   "bits_per_second":      90673.8036234242,
                   "jitter_ms":    38.299352718553145,
                   "lost_packets": 10,
                   "packets":      92,
                   "lost_percent": 10.869565217391305,
                   "sender":       true
           },
           "cpu_utilization_percent":      {
                   "host_total":   3.6519218445696846,
                   "host_user":    0.538746335631838,
                   "host_system":  3.1131657163986644,
                   "remote_total": 0.13084922096488488,
                   "remote_user":  0.033347716674884932,
                   "remote_system":        0.097509783266334346
           }
   }
```

iperf3 server:
```
   ...
   "end":  {
           "streams":      [{
                           "udp":  {
                                   "socket":       5,
                                   "start":        0,
                                   "end":  10.871231,
                                   "seconds":      10.871231,
                                   "bytes":        0,
                                   "bits_per_second":      0,
                                   "jitter_ms":    38.299352718553145,
                                   "lost_packets": 10,
                                   "packets":      92,
                                   "lost_percent": 0,
                                   "out_of_order": 0,
                                   "sender":       false
                           }
                   }],
           "sum":  {
                   "start":        0,
                   "end":  10.871231,
                   "seconds":      10.871231,
                   "bytes":        0,
                   "bits_per_second":      0,
                   "jitter_ms":    38.299352718553145,
                   "lost_packets": 10,
                   "packets":      92,
                   "lost_percent": 10.869565217391305,
                   "sender":       false
           },
           "cpu_utilization_percent":      {
                   "host_total":   0.13084922096488488,
                   "host_user":    0.033347716674884932,
                   "host_system":  0.097509783266334346,
                   "remote_total": 0,
                   "remote_user":  0,
                   "remote_system":        0
           }
   }
}
```

Reply:
SGTM not to use JSON format due to this issue